simulation/traffic_simulator/include/traffic_simulator/hdmap_utils/cache.hpp

```diff
@@ -0,0 +1,136 @@
+#ifndef TRAFFIC_SIMULATOR__HDMAP_UTILS__CACHE_HPP_
+#define TRAFFIC_SIMULATOR__HDMAP_UTILS__CACHE_HPP_
+
+#include <boost/optional.hpp>
+#include <geometry_msgs/msg/point.hpp>
+#include <mutex>
+#include <scenario_simulator_exception/exception.hpp>
+#include <traffic_simulator/math/catmull_rom_spline.hpp>
+#include <unordered_map>
+#include <vector>
+
+namespace hdmap_utils
+{
+class RouteCache
+{
+public:
+  bool exists(std::int64_t from, std::int64_t to)
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    std::pair<std::int64_t, std::int64_t> key;
+    key.first = from;
+    key.second = to;
+    if (data_.find(key) == data_.end()) {
+      return false;
+    }
+    return true;
+  }
+  std::vector<std::int64_t> getRoute(std::int64_t from, std::int64_t to)
+  {
+    if (!exists(from, to)) {
+      THROW_SIMULATION_ERROR(
+        "route from : ", from, " to : ", to, " does not exists on route chache.");
+    }
+    std::lock_guard<std::mutex> lock(mutex_);
+    const auto ret = data_.at({from, to});
+    return ret;
+  }
+  void appendData(std::int64_t from, std::int64_t to, const std::vector<std::int64_t> & route)
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    data_[{from, to}] = route;
+  }
+
+private:
+  std::unordered_map<std::pair<std::int64_t, std::int64_t>, std::vector<std::int64_t>> data_;
+  std::mutex mutex_;
+};
+
+class CenterPointsCache
+{
+public:
+  bool exists(std::int64_t lanelet_id)
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (data_.find(lanelet_id) == data_.end()) {
+      return false;
+    }
+    return true;
+  }
+  std::vector<geometry_msgs::msg::Point> getCenterPoints(std::int64_t lanelet_id)
+  {
+    if (!exists(lanelet_id)) {
+      THROW_SIMULATION_ERROR("center point of : ", lanelet_id, " does not exists on route chache.");
+    }
+    std::lock_guard<std::mutex> lock(mutex_);
+    return data_.at(lanelet_id);
+  }
+  std::shared_ptr<traffic_simulator::math::CatmullRomSpline> getCenterPointsSpline(
+    std::int64_t lanelet_id)
+  {
+    if (!exists(lanelet_id)) {
+      THROW_SIMULATION_ERROR("center point of : ", lanelet_id, " does not exists on route chache.");
+    }
+    std::lock_guard<std::mutex> lock(mutex_);
+    return splines_[lanelet_id];
+  }
+  void appendData(std::int64_t lanelet_id, const std::vector<geometry_msgs::msg::Point> & route)
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    data_[lanelet_id] = route;
+    splines_[lanelet_id] = std::make_shared<traffic_simulator::math::CatmullRomSpline>(route);
+  }
+
+private:
+  std::unordered_map<std::int64_t, std::vector<geometry_msgs::msg::Point>> data_;
+  std::unordered_map<std::int64_t, std::shared_ptr<traffic_simulator::math::CatmullRomSpline>>
+    splines_;
+  std::mutex mutex_;
+};
+
+class LaneletLengthCache
+{
+public:
+  bool exists(std::int64_t lanelet_id)
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (data_.find(lanelet_id) == data_.end()) {
+      return false;
+    }
+    return true;
```

Comment:
This conditional expression is too verbose.
You can simply write `return data_.find(lanelet_id) != data_.end();`.

Reply:
I do not like ternary operator.

Reply:
This is NOT ternary operator.

Reply:
OK, I understand, but it is not readable for me.